java-manta-client/src/main/java/com/joyent/manta/client/crypto/MantaEncryptedObjectInputStream.java

```diff
@@ -94,22 +95,114 @@
+     * @param authenticateCiphertext when true we perform authentication on the ciphertext
      */
     public MantaEncryptedObjectInputStream(final MantaObjectInputStream backingStream,
-                                           final SecretKey secretKey) {
+                                           final SecretKey secretKey,
+                                           final boolean authenticateCiphertext) {
+        this(backingStream, secretKey, authenticateCiphertext, null, null);
+    }
+
+    /**
+     * Creates a new instance that decrypts the backing stream with the specified key.
+     *
+     * @param backingStream stream to read data from
+     * @param secretKey secret key used to decrypt
+     * @param authenticateCiphertext when true we perform authentication on the ciphertext
+     *                               value is ignored when operating with a AEAD cipher mode
+     * @param startPositionInclusive starting position to read plaintext from
+     * @param plaintextRangeLength the total length of plaintext bytes to read
+     */
+    public MantaEncryptedObjectInputStream(final MantaObjectInputStream backingStream,
+                                           final SecretKey secretKey,
+                                           final boolean authenticateCiphertext,
+                                           final Long startPositionInclusive,
+                                           final Long plaintextRangeLength) {
         super(backingStream);
 
+        this.startPosition = startPositionInclusive;
         this.cipherDetails = findCipherDetails();
+
+        if ((startPositionInclusive != null || plaintextRangeLength != null) && !cipherDetails.supportsRandomAccess()) {
+            String msg = "Cipher and cipher mode specified doesn't support random access";
+            MantaClientEncryptionException e = new MantaClientEncryptionException(msg);
+            annotateException(e);
+            throw e;
+        }
+
         this.cipher = cipherDetails.getCipher();
         this.secretKey = secretKey;
         this.hmac = findHmac();
-        initializeCipher();
-        initializeHmac();
+        this.authenticateCiphertext = authenticateCiphertext;
+
+        this.initialBytesToSkip = initializeCipher();
+        this.plaintextRangeLength = verifyAndConditionallyRecalculatePlaintextLength(
+                plaintextRangeLength);
         this.cipherInputStream = createCryptoStream();
+        initializeHmac();
+    }
+
+    /**
+     * Verifies that the plaintext length supplied is less than or equal to the
+     * size of the original plaintext object. If it is sized improperly, we
+     * attempt to convert the plaintext length to the size of the plaintext object.
+     * If we have insufficient information, the plaintext value returned may
+     * be inaccurate, so it is ultimately the responsibility of the caller to
+     * make sure that this value is accurate.
```

Comment:
I imagined a number of cases where the consumer of the SDK would specify a byte range that exceeds the size of the end of the file. I've seen this handled before by just defaulting to the end of the file. It seems rather draconian to insist that the value is correct if we can cope with it. In fact, HTTP range requests operate in this model as well.

The autocorrect is exercised in the following unit tests:
* `MantaClientRangedIT.canGetWithEndRangeBeyondObjectSize`
* `MantaClientRangedIT.canGetWithUnboundedEndRange`

The first one (`MantaClientRangedIT.canGetWithEndRangeBeyondObjectSize`) is an obvious use of it. The second one is interesting. In order to do unbounded ranges, I just set a range to the maximum ciphertext size.

I went and removed the comment about the value being inaccurate because in the `HttpHelper.attemptToFindPlaintextSize()` method we explicitly fail if we are unable to calculate an accurate value.